src/dex/swaap-v1/swaap-v1.ts

```diff
@@ -0,0 +1,633 @@
+  // any pools can be used.
+  getPricesVolume(
+    _from: Token,
+    _to: Token,
+    amounts: bigint[],
+    side: SwapSide,
+    blockNumber: number,
+    limitPools?: string[],
+  ): Promise<null | ExchangePrices<SwaapV1Data>> {
+    return this.getPricesVolumeLogic(
+      _from,
+      _to,
+      amounts,
+      side,
+      blockNumber,
+      null,
+      limitPools,
+    );
+  }
+
+  async getPricesVolumeLogic(
+    _from: Token,
+    _to: Token,
+    amounts: bigint[],
+    side: SwapSide,
+    blockNumber: number,
+    _currentTimeStamp: bigint | null,
+    limitPools?: string[],
+  ): Promise<null | ExchangePrices<SwaapV1Data>> {
+    const from = this.dexHelper.config.wrapETH(_from);
+    const to = this.dexHelper.config.wrapETH(_to);
+
+    const topPools = await this.getTopPoolsForTokens(
+      [from.address, to.address],
+      10,
+      blockNumber,
+    );
+    const allowedPools = limitPools
+      ? topPools.filter((p: SubgraphPoolBase) =>
+          limitPools.includes(`${this.dexKey}_${p.id.toLowerCase()}`),
+        )
+      : topPools;
+
+    if (!allowedPools.length) return null;
+
+    const unitVolume = getBigIntPow(
+      (side === SwapSide.SELL ? from : to).decimals,
+    );
+
+    const srcTokenAddress = from.address.toLowerCase();
+    const destTokenAddress = to.address.toLowerCase();
+
+    if (srcTokenAddress === destTokenAddress) return null;
+    const currentTimestamp =
+      _currentTimeStamp || BigInt(Math.floor(Date.now() / 1000)); // the on-chain pricing logic is not too much sensitive to that value
+    return await Promise.all(
+      allowedPools.map(p =>
+        (async () => {
+          const poolAddress = p.id;
+          let state = this.eventPools[poolAddress].getState(blockNumber);
+          if (!state) {
+            state = await this.eventPools[poolAddress].generateState(
+              blockNumber,
+            );
+          }
+          let [unit, prices] = this.getPricesPool(
+            from.address,
+            to.address,
+            unitVolume,
+            amounts,
+            side,
+            state,
+            currentTimestamp,
+          );
+
+          if (unit === 0n) {
+            // If we didn't fulfill unit amount, scale up the latest amount with a positive price till unit
+            let i = prices.length - 1;
+            let largestAmountPositivePrice = prices.slice(i)[0];
+            while (largestAmountPositivePrice == 0n && i > 0) {
+              --i;
+              largestAmountPositivePrice = prices.slice(i)[0];
+            }
+            if (largestAmountPositivePrice > 0n) {
+              unit =
+                (unitVolume * largestAmountPositivePrice) / amounts.slice(i)[0];
+            }
+          }
+          return {
+            prices,
+            unit,
+            data: {
+              pool: poolAddress,
+            },
+            poolAddresses: [poolAddress],
+            exchange: this.dexKey,
+            gasCost: 610 * 1000,
```

Comment:
`gasCost: 610 * 1000` is it always the case ? Can you also create a global variable in capital case with this value?

`const SWAAPV1_GAS_COST = 610 * 1000;`

Reply:
Actually this number is probably too high.
But generally speaking our gas consumption might double depending on the state of the pool (let’s say from 250k to 500k).
And so we rather prefer to have a constant 500k gas instead of having failure because the state changed between the pricing and the execution (for instance during pricing the gas was 250k but during execution it’s now 500k and so reverts because on 'out-of-gas').
But maybe you have mechanisms to ensure this won’t happen, for instance you might add a margin on the gas already, and in that case we might no need this precaution?

Reply:
Are you able to predict the gas cost during a pricing ?
You can provide an array of gasCost representing the gas cost needed for every elements inside the amounts array. More precise you gas cost estimation are more likely our algorithm will choose your dex. We don't add any marging ourself but you can provide one if your prediction are not perfect :)

Reply:
Hi, the gas costs returned from pricing are just for a rough estimate, it is not used for transaction execution, where we do a proper estimateGas on the whole transaction and add a large margin of error.

If you are able to determine a gas cost from the state of the pool I'd encourage you to do that, otherwise you could just split the difference and use something like 375k.  You are not doing yourself any favours by setting the value too high, it will just make your DEX less likely to be chosen.